// @notifee/react-native/dist/validators/validateIOSPermissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
function validateIOSPermissions(permissions) {
    const out = {
        alert: true,
        badge: true,
        sound: true,
        carPlay: true,
        settings: true,
        provisional: false,
        announcement: false,
    };
    if (!permissions) {
        return out;
    }
    if (utils_1.hasOwnProperty(permissions, 'alert')) {
        if (!utils_1.isBoolean(permissions.alert)) {
            throw new Error("'alert' expected a boolean value.");
        }
        out.alert = permissions.alert;
    }
    if (utils_1.hasOwnProperty(permissions, 'badge')) {
        if (!utils_1.isBoolean(permissions.badge)) {
            throw new Error("'alert' badge a boolean value.");
        }
        out.badge = permissions.badge;
    }
    if (utils_1.hasOwnProperty(permissions, 'sound')) {
        if (!utils_1.isBoolean(permissions.sound)) {
            throw new Error("'sound' expected a boolean value.");
        }
        out.sound = permissions.sound;
    }
    if (utils_1.hasOwnProperty(permissions, 'carPlay')) {
        if (!utils_1.isBoolean(permissions.carPlay)) {
            throw new Error("'carPlay' expected a boolean value.");
        }
        out.carPlay = permissions.carPlay;
    }
    if (utils_1.hasOwnProperty(permissions, 'settings')) {
        if (!utils_1.isBoolean(permissions.settings)) {
            throw new Error("'settings' expected a boolean value.");
        }
        out.settings = permissions.settings;
    }
    if (utils_1.hasOwnProperty(permissions, 'provisional')) {
        if (!utils_1.isBoolean(permissions.provisional)) {
            throw new Error("'provisional' expected a boolean value.");
        }
        out.provisional = permissions.provisional;
    }
    if (utils_1.hasOwnProperty(permissions, 'announcement')) {
        if (!utils_1.isBoolean(permissions.announcement)) {
            throw new Error("'announcement' expected a boolean value.");
        }
        out.announcement = permissions.announcement;
    }
    return out;
}
exports.default = validateIOSPermissions;
